src/Compiler/Microsoft.NET.Sdk.Razor.SourceGenerators/RazorSourceGenerator.cs

```diff
@@ -174,40 +160,40 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
                     {
                         // If there's no razor code in this app, don't do anything.
                         RazorSourceGeneratorEventSource.Log.DiscoverTagHelpersFromReferencesStop();
-                        return ImmutableArray<TagHelperDescriptor>.Empty;
+                        return null;
+                    }
+
+                    // If there aren't any references, we can bail out early.
+                    // Note: compilation.References just enuemrators compilation.ExternalReferences and compilation.DirectiveReferences.
+                    if (compilation.ExternalReferences is [] && compilation.DirectiveReferences is [])
+                    {
+                        return null;
```

Comment:
Perhaps `Debug.Assert(!compilation.References.Any())`

Reply:
Is that a concerning case? Are there scenarios where a source generator might run before all of the references have been added to a compilation by the project system?

Reply:
I just meant to assert that inside the `if`. Verify that if `compilation.ExternalReferences is [] && compilation.DirectiveReferences is []` is `true` than also `!compilation.References.Any()` is `true` because then below we iterate over `compilation.References`. But actually there will be no error if the assumption doesn't hold, so no need to add it.

Reply:
Thanks! I understand now.